src/pipeline/utilities/SpatialLocationCalculator/SpatialUtils.cpp

```diff
@@ -0,0 +1,408 @@
+        if(static_cast<int>(segmentationMaskHeight) != depthHeight || static_cast<int>(segmentationMaskWidth) != depthWidth) {
+            logger->debug("Segmentation mask size {}x{} does not match depth frame size {}x{}. Resizing to depth frame size.", segmentationMaskWidth, segmentationMaskHeight, depthWidth, depthHeight);
+            cv::resize(seg, seg, cv::Size(depthWidth, depthHeight), 0, 0, cv::INTER_NEAREST);
+        }
+        maskPtr = seg.data;
+    }
+
+    const uint16_t* plane = (uint16_t*)depthFrame->data->getData().data();
+
+    for(int i = 0; i < static_cast<int>(imgDetectionsVector.size()); i++) {
+        const dai::ImgDetection& detection = imgDetectionsVector[i];
+        dai::RotatedRect rotatedRect = detection.getBoundingBox();
+        rotatedRect = rotatedRect.denormalize(depthWidth, depthHeight, true);
+
+        const auto& outerPoints = rotatedRect.getOuterRect();
+        unsigned int xstart = std::min(std::max(0, (int)outerPoints[0]), depthWidth - 1);
+        unsigned int ystart = std::min(std::max(0, (int)outerPoints[1]), depthHeight - 1);
+        unsigned int xend = std::min(depthWidth - 1, (int)outerPoints[2]);
+        unsigned int yend = std::min(depthHeight - 1, (int)outerPoints[3]);
+
+        std::uint32_t sum = 0;
+        std::uint16_t min = 65535, max = 0;
+        std::uint32_t counter = 0;
+
+        std::vector<uint16_t> validPixels;
+        for(unsigned int y = ystart; y < yend; y += stepSize) {
+            for(unsigned int x = xstart; x < xend; x += stepSize) {
+                int index = y * depthWidth + x;
+                uint16_t px = plane[index];
+
+                bool usePixel = true;
+                if(useSegmentation) {
+                    usePixel = (maskPtr) ? (maskPtr[index] == i) : true;
+                }
+
+                if(px > lowerThreshold && px < upperThreshold && usePixel) {
+                    validPixels.push_back(px);
+                    sum += px;
+                    ++counter;
+                    if(px < min) min = px;
+                    if(px > max) max = px;
+                }
+            }
+        }
+        dai::SpatialImgDetection spatialDetection;
+        if(calculateSpatialKeypoints) {
+            logger->info("Calculating spatial keypoints for detection {}", i);
+            std::vector<dai::SpatialKeypoint> spatialKeypoints;
+            for(auto kp : detection.getKeypoints()) {
+                int kpx = static_cast<int>(kp.imageCoordinates.x * depthWidth);
+                int kpy = static_cast<int>(kp.imageCoordinates.y * depthHeight);
+
+                int kpxstart = std::max(0, kpx - keypointRadius);
+                int kpystart = std::max(0, kpy - keypointRadius);
+                int kpxend = std::min(depthWidth - 1, kpx + keypointRadius);
+                int kpyend = std::min(depthHeight - 1, kpy + keypointRadius);
+
+                std::uint32_t kpSum = 0;
+                std::uint16_t kpMin = 65535, max = 0;
```

Comment:
Just a nit pick, it's rather clear that 65535 is the maximum representable number using 16 bits, I'd make it a little more explicit though. 
```suggestion
                std::uint16_t kpMin = std::numeric_limits<uint16_t>::max(), max = std::numeric_limits<uint16_t>::min();
```

Also, is the naming here correct? There is another declaration of `max` a few lines above and it's not clear which one is intended to be used when
`float z = calculateDepth(calculationAlgorithm, validPixels, sum, counter, min, max);` happens

Reply:
Thanks for catching!